### PR TITLE
Feature/login

### PR DESCRIPTION
**Pull request by msdaungsangi**

### CODE REVIEW

**Comment by msdaungsangi** [comment]
https://docs.google.com/spreadsheets/d/16-A1yjOXiQPoTwT-kNkS-5MpvUoznBai-nwQy1y2HFE/edit#gid=1308544392

File: app/Contracts/Dao/LoginDaoInterface.php [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
+namespace App\Contracts\Dao;
+use Illuminate\Http\Request;
+
+interface LoginDaoInterface{
+  public function updatePassword(Request $request, array $password);
```
Comment (ingyinmaycgm) [inline app/Contracts/Dao/LoginDaoInterface.php:6]: Use four-space indentation.

File: app/Contracts/Services/LoginServiceInterface.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?php
+namespace App\Contracts\Services;
+use Illuminate\Http\Request;
+
+interface LoginServiceInterface{
+  public function updatePassword(Request $request, array $password);
```
Comment (ingyinmaycgm) [inline app/Contracts/Services/LoginServiceInterface.php:6]: Same here . 

> Use four-space indentation.

Check all php file and fix indentation unless 4-space indentation.

File: app/Dao/LoginDao.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace App\Dao;
+
+use App\Contracts\Dao\LoginDaoInterface;
+use Illuminate\Http\Request;
+
+class LoginDao implements LoginDaoInterface
+{
+    public function updatePassword(Request $request, array $password)
```
Comment (ingyinmaycgm) [inline app/Dao/LoginDao.php:10]: Add function doc.

File: app/Http/Controllers/Auth/LoginController.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    /**
+     * Create a new controller instance.
+     *
+     * @return void
+     */
+    public function __construct()
+    {
+        $this->middleware('guest')->except('logout');
+    }
+
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/Auth/LoginController.php:43]: Add function doc. 
Check the whole file.

File: app/Http/Controllers/Auth/LoginController.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        $getUser = $request->validate([
+            'email' => 'required',
+            'password' => 'required',
+        ]);
+
+        if (Auth::attempt($getUser)) {
+            return redirect()->intended('home');
+        }
+        
+        return back()->with('error', 'Invalid email or password');
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/Auth/LoginController.php:60]: Use constant messages.

File: app/Http/Controllers/Auth/RegisterController.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * Create a new user instance after a valid registration.
+     *
+     * @param  array  $data
+     * @return \App\Models\User
+     */
+    protected function create(array $data)
+    {
+        return User::create([
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/Auth/RegisterController.php:67]: Do this in DAO. 
Did you forget to check PR check list?

File: app/Http/Controllers/HomeController.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    public function updatePassword(Request $request)
+    {
+        $request->validate([
+            'old_password' => 'required',
+            'new_password' => 'required|max:255',
+            'new_password_confirmation' => 'required|max:255',
+        ]);
+        $samePass = config('messages.login.pass_same');
+        $oldPass = config('messages.login.old_pass_incorrect');
+        $success = config('messages.login.pass_change_success');
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/HomeController.php:53]: Waste of memory. 
Declare it only if it's used.
May be not all 3 vars are used at the same time.

File: app/Services/LoginService.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+use App\Contracts\Dao\LoginDaoInterface;
+use App\Contracts\Services\LoginServiceInterface;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Hash;
+
+class LoginService implements LoginServiceInterface
+{
+    protected $loginDao;
+
+    public function __construct(LoginDaoInterface $loginDao)
```
Comment (ingyinmaycgm) [inline app/Services/LoginService.php:14]: Add function doc for the whole file.

File: app/Services/UserService.php [hunk 9]
```diff
@@ -50,29 +51,55 @@ public function getUsers()
...
+            }
+            $data = [
+                'email' => $data['email'],
+                'name' => $data['name'],
+                'password' => Hash::make($data['password']),
+                'img' => $imageName,
+                'role' => $data['role'],
+            ];
+
+            $this->userDao->createUser($request, $data);
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php:99]: There code lines seem to be the same for both condition right? 
Except `created_by`.
Try to use as common code.

File: app/Services/UserService.php [hunk 10]
```diff
@@ -95,26 +121,51 @@ public function getUserById(int $id)
...
+            }
+            $data = [
+                'name' => $data['name'],
+                'email' => $data['email'],
+                'password' => Hash::make($data['password']),
+                'img' => $imageName,
+                'role' => $data['role'],
+            ];
+
+            $this->userDao->updateUser($request, $data, $id);
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php:167]: Same here. 

> There code lines seem to be the same for both condition right?
> Except created_by.
> Try to use as common code.

File: app/Dao/LoginDao.php [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+use App\Contracts\Dao\LoginDaoInterface;
+use Illuminate\Http\Request;
+
+/**
+ * LoginDao
+ */
+class LoginDao implements LoginDaoInterface
+{
+    public function updatePassword(Request $request, array $password)
```
Comment (ingyinmaycgm) [inline app/Dao/LoginDao.php:13]: How about this? 
https://github.com/msdaungsangi/project-laravel/pull/3/files#r1348208282

File: app/Http/Controllers/UserController.php [hunk 12]
```diff
@@ -58,6 +58,19 @@ public function create()
         return view('users.create');
     }
 
+    public function registUser(Request $request)
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/UserController.php:61]: Add function doc.
Check the whole file and add.

File: app/Services/UserService.php [hunk 13]
```diff
@@ -61,18 +83,19 @@ public function createUser(Request $request, array $data)
...
             $imageName = null;
         }
+
         $data = [
             'email' => $data['email'],
             'name' => $data['name'],
             'password' => Hash::make($data['password']),
             'img' => $imageName,
             'role' => $data['role'],
+            'created_by' => $createdBy,
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php:93]: If  `Auth::check()` fails, then this code line will raise error. (undefined var)
Make testing and fix.

File: app/Services/UserService.php [hunk 14]
```diff
@@ -112,8 +138,9 @@ public function updateUser(Request $request, array $data, int $id)
             'password' => Hash::make($data['password']),
             'img' => $imageName,
             'role' => $data['role'],
+            'updated_by' => $updatedBy,
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php:141]: Same here. 

> If Auth::check() fails, then this code line will raise error. (undefined var)
> Make testing and fix.

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 15]
```diff
@@ -17,7 +17,7 @@ public function up(): void
             $table->string('password');
             $table->string('name');
             $table->string('img')->nullable();
-            $table->integer('role');
+            $table->integer('role')->default('2');
```
Comment (ingyinmaycgm) [inline database/migrations/2014_10_12_000000_create_users_table.php:20]: Please don't add default value for role. 
Make validation for input request and insert user input to table.

**Pull request merged**